app.rb

```diff
@@ -175,6 +175,69 @@ def respond_with(format, category, responses)
     end
   end
 
+  def form_xml_for_settings(category, tag, responses)
+    builder = Nokogiri::XML::Builder.new do |xml|
+      xml.send(category) {
+        responses.each do |response|
+          case category
+          when :providers
+             xml.provider(tag => response)
+          when :servers
+             xml.server(tag => response)
+          when :images
+             xml.image(tag => response)
+          when :regions
+             xml.region(tag => response)
+          end
+        end
+      }
+    end
+    return builder
+  end
+
+  get '/v1/providers.?:ext?' do
+    validate_params_ext
+    
+    settings_xml = form_xml_for_settings(:providers, :name, settings.providers)
+    responses = settings_xml.doc.css("providers>provider")
```

Comment:
Instead of sending the parent tag, using a case to set the child tag, then filtering out the children, please consider explicitly sending the child tag, and in the processor, setting a generic parent, and filtering it off; e.g.

```
  def form_xml_for_settings(tag, values)
    Nokogiri::XML::Builder.new do |xml|
      xml.root {
        values.each { |value| xml.send(tag, value) }
      }
    end.doc.css(tag)
  end

  get '/v1/states.?:ext?' do
    validate_params_ext
    responses = form_xml_for_settings('state', settings.image_states)
    respond_with params[:ext], :states, responses
  end
```